File: R/forLater/crop.R
#' Interactive data cropping tool.
#' 
#' This function plots the input data # and allows the user to select start and end times for cropping.
#' 
#' Possible input combinations include: crop(X) if X is a sensor list, crop(X, sampling_rate) if X is a vector or matrix.
#' @param X A sensor list, vector or matrix. X can be regularly or irregularly sampled data in any frame and unit.
#' @param sampling_rate The sampling rate of X in Hz. This is only needed if X is not a sensor list. If X is regularly sampled, sampling_rate is one number. sampling_rate may also be a vector of sampling times for X. This is only needed if X is not a sensor list and X is not regularly sampled.
#' @return A list with 3 elements:
#' \itemize{
#'  \item{\strong{Y: }} A sensor list, vector or matrix containing the cropped data segment. If the input is a sensor list, the output will also be. The output has the same units, frame and sampling characteristics as the input.
#'  \item{\strong{T: }} A vector of sampling times for Y. This is only returned if X is irregularly sampled and X is not a sensor list. If X is a sensor list, the sampling times are stored in the list.
#'  \item{\strong{tcues: }} tcues is a two-element vector containing the start and end time cue in seconds of the data segment kept, i.e., tcues = c(start_time, end_time).
#' }
#' @example load_nc('testset3')
#'          Pc <- crop(P)		#interactively select a section of data
#'          plott(Pc)
#'          #plot shows the cropped section

crop <- function(X, sampling_rate) {
  if (missing(X)) {
    stop("X is a required input")
  }
  if (is.list(X)) {
    x <- list$data
    sampling_rate <- X$sampling_rate
    if (is.null(x)) {
      stop("data cannot be empty")
    }
  } else {
    if (missing(sampling_rate)) {
      stop("inputs for X and sampling_rate are required")
    }
    x <- X
    if (is.vector(x) | nrow(x) == 1) {
      x <- matrix(x, nrow = length(x), ncol = 1)
    }
  }
  graphics.off()
  if (length(sampling_rate) > 1) {
    plot(sampling_rate, x, type = "p", xlab = "Time (seconds)")
  } else {
    plot((c(1:nrow(x)) / sampling_rate), x, type = "l", xlab = "Time (seconds)")
  }
  print("Position your cursor and then click once followed by clicking FINISH to change the start, or click twice in the same spot followed by clicking FINISH to change the end")
  if (length(sampling_rate) > 1) {
    tcues <- c(min(sampling_rate), max(sampling_rate))
  } else {
    tcues <- c(0, (nrow(x) / sampling_rate))
  }
  LIMS <- tcues
  lines(matrix((c(1,1) * tcues[1]), ncol = 1), col = "green", lwd = 1.5)
  points(tcues[1], mean((par("usr")[3]+.04), (par("usr")[4]-.04)), col = "green", pch = 6, cex = 12)
  lines(matrix((c(1,1) * tcues[2]), ncol = 1), col = "red", lwd = 1.5)
  points(tcues[2], mean((par("usr")[3]+.04), (par("usr")[4]-.04)), col = "red", pch = 6, cex = 12)
  pts <- graphics::locator(2)
  if (length(pts$x) == 1) {
    t <- pts$x
    tcues[1] <- max(LIMS[1], min(t, tcues[2]))
    lines(matrix((c(1,1) * tcues[1]), nrow = 1), col = "green", lwd = 1.5)
    points(tcues[1], mean((par("usr")[3]+.04), (par("usr")[4]-.04)), col = "green", pch = 6, cex = 12)
    lines(matrix((c(1,1) * tcues[2]), ncol = 1), col = "red", lwd = 1.5)
    points(tcues[2], mean((par("usr")[3]+.04), (par("usr")[4]-.04)), col = "red", pch = 6, cex = 12)
  } else {
    t < pts$x[2]
    tcues[2] <- max(LIMS[2], min(t, tcues[1]))
    lines(matrix((c(1,1) * tcues[1]), ncol = 1), col = "green", lwd = 1.5)
    points(tcues[1], mean((par("usr")[3]+.04), (par("usr")[4]-.04)), col = "green", pch = 6, cex = 12)
    lines(matrix((c(1,1) * tcues[2]), nrow = 1), col = "red", lwd = 1.5)
    points(tcues[2], mean((par("usr")[3]+.04), (par("usr")[4]-.04)), col = "red", pch = 6, cex = 12)
  }
  if (is.list(X)) {
    Y <- crop_to(X, tcues)$X
    T <- tcues
  } else {
    list <- crop_to(x, sampling_rate, tcues)
    Y <- list$X
    T <- list$T
    if (length(sampling_rate) == 1) {
      T <- tcues
    }
  }
  return(list(Y = Y, T = T, tcues = tcues))
}

File: matlab/calibrate/crop.m
function    [Y,T,tcues] = crop(X,fs)

%		Y = crop(X)				% X is a sensor structure
%		or
%		[Y,tcues] = crop(X,fs)			% X is a regularly sampled vector or matrix
%		or
%		[Y,T,tcues] = crop(X,T)			% X is a regularly sampled vector or matrix
%
%		Interactive data cropping tool. This function plots the input data #
%		and allows the user to select start and end times for cropping.
%
%     Inputs:
%     X is a sensor structure, vector or matrix. X can be regularly or
%		 irregularly sampled data in any frame and unit.
%     fs is the sampling rate of X in Hz. This is only needed if
%		 X is not a sensor structure and X is regularly sampled.
%		T is a vector of sampling times for X. This is only needed if X is
%		 not a sensor structure and X is not regularly sampled.
%
%     Results:
%     Y is a sensor structure, vector or matrix containing the cropped data segment.
%		 If the input is a sensor structure, the output will also be. The output has
%		 the same units, frame and sampling characteristics as the input.
%     T is a vector of sampling times for Y. This is only returned if X is irregularly
%	    sampled and X is not a sensor structure. If X is a sensor structure, the sampling
%		 times are stored in the structure.
%		tcues is a two-element vector containing the start and end time cue
%		 in seconds of the data segment kept, i.e., tcues = [start_time, end_time].
%
%		Example:
%		 loadnc('testset3')
%		 Pc = crop(P);		% interactively select a section of data
%		 plott(Pc)
%		 % plot shows the cropped section
%
%     Valid: Matlab, Octave
%     markjohnson@st-andrews.ac.uk
%     Last modified: 28 July 2017

T = [] ;
if nargin<1,
	help crop
	return
end
	
if isstruct(X),
	[x,fs] = sens2var(X) ;
	if isempty(x), return, end
else
	if nargin<2,
		help crop
		return
	end
	x = X ;
	if size(x,1)==1,		% make sure x is a column vector
		x = x(:) ;
	end
end

clf					% clear the current figure
if length(fs)>1,
	plot(fs,x,'.') ;
else
	plot((1:size(x,1))/fs,x) ;
end
xlabel('Time (seconds)') ;
ht = title('Position the cursor and type s or e to adjust start or end. Type q to finish') ;
set(ht,'color','r','fontsize',12)

zoom off
if length(fs)>1,
	tcues = [min(T) max(T)] ;
else
	tcues = [0 size(x,1)/fs] ;
end

LIMS = tcues ;
hold on
hs = plot([1;1]*tcues(1),get(gca,'YLim'),'g') ;
set(hs,'LineWidth',1.5) ;
hsm = plot(tcues(1),mean(get(gca,'YLim')),'g>') ;
set(hsm,'MarkerSize',12,'MarkerFaceColor','g') ;

he = plot([1;1]*tcues(2),get(gca,'YLim'),'r') ;
set(he,'LineWidth',1.5) ;
hem = plot(tcues(2),mean(get(gca,'YLim')),'r<') ;
set(hem,'MarkerSize',12,'MarkerFaceColor','r') ;

while 1,
   [t,y,s] = ginput(1) ;
	if isempty(t), break, end
   s = char(s) ;
   switch char(s),
      case 's'
         tcues(1) = max(LIMS(1),min(t,tcues(2))) ;
         set(hs,'XData',tcues(1)*[1 1]) ;
         set(hsm,'XData',tcues(1)) ;
      case 'e'
         tcues(2) = min(LIMS(2),max(t,tcues(1))) ;
         set(he,'XData',tcues(2)*[1 1]) ;
         set(hem,'XData',tcues(2)) ;
		case 'q'
			break
   end
end

hold off
if isstruct(X),
	Y = crop_to(X,tcues) ;
	T = tcues ;
else
	[Y,T] = crop_to(x,fs,tcues) ;
	if length(fs)==1,
		T = tcues ;
	end
end
return
